iceoryx_hoofs/include/iceoryx_hoofs/error_handling_2/platform/test_platform/polymorphic_handler.hpp

```diff
@@ -0,0 +1,120 @@
+#pragma once
+
+#include <atomic>
+#include <mutex>
+#include <type_traits>
+
+namespace eh
+{
+// we lose generality now wrt. the interface (require activate, deactivate etc. in the interface)
+template <typename Interface, typename Default>
+class PolymorphicHandler
+{
+    static_assert(std::is_base_of<Interface, Default>::value, "Default must inherit from Interface");
+
+  public:
+    using Self = PolymorphicHandler<Interface, Default>;
+
+    // on first call (in a thread):
+    // 1. local is initialized
+    //    - getCurrent is called
+    //    - instantiates singleton instance()
+    //    - instantiates default
+    //    - sets m_current of instance to default instance (release)
+    //    - default is active
+
+    // if any thread changes the active handler with set (or reset)
+    // under mutex protection, it will:
+    //    - set the new handler to active
+    //    - set current handler to the new handler
+    //    - deactivate the old handler (can still be used as it still needs to exist)
+    //
+
+    // on any call after the handler was changes in another thread
+    // 1. we check whether the handler is active
+    // (can be outdated information but will eventually be false once the atomic value is updated)
+    // 2. if it was changed it is now inactive and we wait to obtain the mutex lock
+    // under lock, we update the local handler to the new one (note that it cannot change
+    // while this happens as we hold the lock)
+
+    static Interface& get()
+    {
+        thread_local Interface* local = getCurrent(); // initialized once per thread on first call
+
+        if (!local->isActive())
+        {
+            std::lock_guard<std::mutex> lock(instance().m_mutex);
+            local = getCurrent();
+        }
+        return *local;
+    }
+
+    static Interface* set(Interface& handler)
+    {
+        auto& i = instance();
+
+        // setting is rare and we lock to synchronize and update the active flags properly
+        std::lock_guard<std::mutex> lock(i.m_mutex);
+        if (i.m_isFinal)
+        {
+            std::cerr << "SETTING THE HANDLER AFTER FINALIZE IS NOT ALLOWED!" << std::endl;
+            std::terminate();
+            return nullptr;
+        }
+
+        handler.activate(); // it may have been deactivated before, so always reactivate it
+        auto prev = i.m_current.exchange(&handler, std::memory_order_relaxed);
+
+        // anyone still using it will sooner or later see that it is inactive
+        // and switch to the new one
+        prev->deactivate();
+        return prev;
+    }
+
+    static Interface* reset()
+    {
+        return set(getDefault());
+    }
+
+    static void finalize()
+    {
+        auto& s = instance();
+        std::lock_guard<std::mutex> lock(s.m_mutex);
+        s.m_isFinal.store(true);
+    }
+
+  private:
+    std::atomic_bool m_isFinal{false};
+    std::mutex m_mutex;
+    std::atomic<Interface*> m_current;
```

Comment:
I know, I wanted to avoid the mutex on the fast path (it will not be contended but I am fairly sure it is not as fast as a bare atomic, but close if not contended). Not that interesting for an error handler though as it is rare.

Reply:
Well, I slightly changed my mind. It's choosing between two evils. Usually there should be no contention but there is still a mutex involved and we might have to ensure that it does not block indefinitely. Whereas with your proposal, there is the issue with lock-free code. It is so simple that it should be safe. Therefore I'm now slightly leaning towards the atomic pointer